### PR TITLE
chore: implement `@puppeteer/doctest`

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR implements a documentation example tester, similar to the [one in Rust](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html).

## Usage

```
npm run doctest
```

## Details

`@puppeteer/doctest` looks through compiled code for `@example` comments with fenced code blocks. It will then parse out the code within the fence and run it using Node's test runner.

The first line containing the starting fence defines the options for a given fenced code block. For example,

 - <code>```ts</code> defines a TypeScript code block.
 - <code>```</code> defines a JavaScript code block.
 - <code>```ts ignore</code> defines a TypeScript code block that is ignored for testing.

Note that

 - <code>``` ignore</code> and 
 - <code>```ignore</code>

are different. The former is an ignored JS code block. The latter is a code block for the `ignore` language.

If there are several code blocks within an example, they are concatenated before running.

### CODE REVIEW

**Comment by OrKoN** [comment]
I think it is a good idea, I will let @Lightning00Blade to do the review

File: test/TestExpectations.json [hunk 1]
```diff
@@ -779,6 +779,12 @@
     "parameters": ["webDriverBiDi"],
     "expectations": ["SKIP"]
   },
+  {
+    "testIdPattern": "[launcher.spec] Launcher specs Puppeteer Puppeteer.launch should launch Chrome properly with --no-startup-window and waitForInitialPage=false",
```
Comment (OrKoN) [inline test/TestExpectations.json]: let's investigate flakiness today and address it separately

Comment (OrKoN) [reply test/TestExpectations.json]: never mind (will post on chat)

File: tools/doctest/package.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "@puppeteer/doctest",
+  "version": "0.1.0",
+  "type": "module",
+  "bin": "./bin/doctest.js",
+  "description": "Tests JSDoc @example code within a file.",
```
Comment (Lightning00Blade) [inline tools/doctest/package.json:7]: Nit:
```suggestion
  "description": "Tests JSDoc/TSDoc @example code within a file.",
```

File: tools/doctest/src/doctest.ts [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    type: 'string',
+  })
+  .array('files')
+  .version(false)
+  .help()
+  .parse();
+
+for (const pattern of files) {
+  for await (const file of new Glob(pattern, {})) {
+    await test(file, async context => {
```
Comment (Lightning00Blade) [inline tools/doctest/src/doctest.ts]: Nit: lets use `it` its more consistent with what we use else where.
Also maybe wrap this in Describe to be more obvious what that we are using Node:test
To reduce nesting and complexity I would extract the `test` callback to helper.

File: tools/doctest/src/doctest.ts [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+      if (!description) {
+        continue;
+      }
+      const lines = description.split('\n');
+      const blocks = [];
+      let context: Context | undefined;
+      for (let i = 0; i < lines.length; i++) {
+        const line = lines[i]!;
+        const borderIndex = line.indexOf(CODE_FENCE);
+        if (borderIndex !== -1) {
```
Comment (Lightning00Blade) [inline tools/doctest/src/doctest.ts]: Inverse this into a continue else the code becomes super nested and hard to read.

File: tools/doctest/src/doctest.ts [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+          if (!context) {
+            const [tag, ...options] = line
+              .slice(borderIndex + CODE_FENCE.length)
+              .split(' ');
+            if (options.includes('ignore')) {
```
Comment (Lightning00Blade) [inline tools/doctest/src/doctest.ts]: Nit: extract to const.

File: tools/doctest/src/doctest.ts [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+              // Ignore the code sample.
+              continue;
+            }
+            const fail = options.includes('fail');
+            if (!tag || tag.match(/js|javascript/)) {
```
Comment (Lightning00Blade) [inline tools/doctest/src/doctest.ts]: I think we should just brake on undefined language. 
I would be more keen on having an EsLint rule to enforce that rather then doing it here.

File: tools/doctest/src/doctest.ts [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+              context = {language: Language.JavaScript, fail, start: i + 1};
```
Comment (Lightning00Blade) [inline tools/doctest/src/doctest.ts]: Add a comment on why the `i + 1` it seems hard to follow. (I presume it source map related)

File: tools/doctest/package.json (hunk 2 above)
Comment (jrandolf) [reply tools/doctest/package.json:7]: It doesn't work with TS files, only JS files.

File: tools/doctest/src/doctest.ts (hunk 3 above)
Comment (jrandolf) [reply tools/doctest/src/doctest.ts]: I already tried this. We can't because `it` has a different type for some reason in `@types/node`...

File: tools/doctest/src/doctest.ts (hunk 6 above)
Comment (jrandolf) [reply tools/doctest/src/doctest.ts]: I think it's reasonable to assume code blocks are in JS since it's in JS.

File: tools/doctest/src/doctest.ts (hunk 7 above)
Comment (jrandolf) [reply tools/doctest/src/doctest.ts]: `i+1` is the starting line for the code block (nothing to do with the source map). I wouldn't say adding a comment would help here; if you get to this point, you need to know where the indexes line up.

File: package.json [hunk 8]
```diff
@@ -59,6 +60,12 @@
         "./packages/puppeteer-core:build:docs"
       ]
     },
+    "doctest": {
+      "command": "npx @puppeteer/doctest 'packages/puppeteer-core/lib/esm/**/*.js'",
+      "dependencies": [
+        "./packages/puppeteer-core:build"
```
Comment (Lightning00Blade) [inline package.json]: Don't you want to make sure the package is build?

File: tools/doctest/package.json (hunk 2 above)
Comment (Lightning00Blade) [reply tools/doctest/package.json:7]: It's not about files you are pointing out JSDoc which does not include TSDoc, which is an extension on top of JSDoc.

File: tools/doctest/src/doctest.ts (hunk 3 above)
Comment (Lightning00Blade) [reply tools/doctest/src/doctest.ts]: Yeah I think you need to warp it inside describe for it to work. Also comment on the other parts as well please.

File: tools/doctest/src/doctest.ts [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+  return generator.toJSON();
+}
+
+async function compileCode(language: Language, code: string) {
+  return (await transform(code, {
+    sourceMaps: true,
+    inlineSourcesContent: false,
+    jsc: {
+      parser: {
+        syntax: (
```
Comment (Lightning00Blade) [inline tools/doctest/src/doctest.ts]: Extract into a variable and then reference it.

File: tools/doctest/src/doctest.ts [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+              {
+                ...position,
+                extracted:
+                  context.code.split('\n').length +
+                  context.positions.at(-1)!.extracted -
+                  // We subtract this because of the accumulated '\n'.
+                  (index - 1),
+              },
+            ],
+            fail: context.fail || fail,
```
Comment (Lightning00Blade) [inline tools/doctest/src/doctest.ts]: Doesn't this imply that if we have 2 blocks on the same comment they can't be ` typescript ` and ` typescript fail` as both will be expected to fail?

File: package.json (hunk 8 above)
Comment (jrandolf) [reply package.json]: It's built during `npm i`

File: tools/doctest/src/doctest.ts (hunk 9 above)
Comment (jrandolf) [reply tools/doctest/src/doctest.ts]: Why? It's only used in one place.

File: tools/doctest/package.json (hunk 2 above)
Comment (jrandolf) [reply tools/doctest/package.json:7]: Considering the parser (doctrine) doesn't really support TSDoc, I think we should just advocate JSDoc.

File: tools/doctest/src/doctest.ts (hunk 10 above)
Comment (jrandolf) [reply tools/doctest/src/doctest.ts]: Since the blocks get concatenated, the code will fail; so yes.

File: package.json (hunk 8 above)
Comment (Lightning00Blade) [reply package.json]: Two things to note:
1) `--ignore-scripts` - sometimes I use it.
2) If i change the package itself and run this command it will use the older version, and I think the consensus on the topic is we want it to rebuild in necessary like other test or tools.

File: tools/doctest/src/doctest.ts (hunk 9 above)
Comment (Lightning00Blade) [reply tools/doctest/src/doctest.ts]: So the idea is you are using it as a map then getting a single value, having a variable makes it easy to understand intention to do.

File: package.json (hunk 8 above)
Comment (jrandolf) [reply package.json]: The problem with `npx` is that it only works if the binary is symlinked in `node_modules/.bin`. This implies if the package was not built, then `npm` has nothing to symlink. We could just do `tools/doctest/bin/doctest.js`, but this creates a dependency on the file name which is what I am preventing.

Open to suggestions.

Comment (jrandolf) [reply package.json]: Solved.

File: tools/doctest/src/doctest.ts (hunk 10 above)
Comment (Lightning00Blade) [reply tools/doctest/src/doctest.ts]: Just to be entirely sure (`.` for ` ):
```
@example
...ts
...

...ts fail
...
``` 
But 
```
@example
...ts
...
@example
...ts fail
...
``` 
work right?

File: tools/doctest/src/doctest.ts (hunk 9 above)
Comment (jrandolf) [reply tools/doctest/src/doctest.ts]: I misread your comment. I thought you meant the entire option (not the map). GitHub doesn't show the map in the preview.

File: tools/doctest/src/doctest.ts (hunk 10 above)
Comment (jrandolf) [reply tools/doctest/src/doctest.ts]: Indeed.

File: tools/doctest/src/doctest.ts (hunk 5 above)
Comment (Lightning00Blade) [reply tools/doctest/src/doctest.ts]: This was resolve but no action or comment left.

**Review by Lightning00Blade — APPROVED** [review]

File: tools/doctest/src/doctest.ts (hunk 5 above)
Comment (jrandolf) [reply tools/doctest/src/doctest.ts]: As discussed offline, we will just keep it as is.
